### PR TITLE
Basic case management

### PR DESCRIPTION
**Pull request by noliveleger**
## Checklist

1. [x] If you've added code that should be tested, add tests
2. [x] If you've changed APIs, update (or create!) the documentation
3. [x] Ensure the tests pass
4. [x] Make sure that your code lints and that you've followed [our coding style](https://github.com/kobotoolbox/kpi/blob/master/CONTRIBUTING.md)
5. [x] Write a description of your work suitable for publishing on [our forum](https://community.kobotoolbox.org/tag/release-notes)
6. [x] Mention any related issues in this repository (as #ISSUE) and in other repositories (as kobotoolbox/other#ISSUE)

## Description
Adds support for KPI paired data represented as `xml-external` in form definition. (see https://xlsform.org/en/#external-xml-data for more detail).
Hash of each `xml-external` is void after 5 minutes (the setting is customizable in configuration file, see `PAIRED_DATA_EXPIRATION`) to force Enketo to download new data

## Related issues
Part of kobotoolbox/kpi#2767

### CODE REVIEW

File: onadata/libs/utils/hash.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+                # With `stream=True`, the connection is kept alive until it is
+                # closed or all the data is consumed. Because, we do not consume
+                # the data, we have to close the connexion manually.
+                response.close()
+                return _prefix_hash(hex_digest=url_hash)
+
+            file_size = int(response.headers['Content-Length'])
+            # If the remote file is smaller than the threshold or smaller than
+            # 3 times the chunk, we retrieve the whole file to avoid extra
+            # requests.
```
Comment (jnm) [inline onadata/libs/utils/hash.py:75]: Let's note in the `{algorithm}` when we use this approach

**Review by jnm — CHANGES_REQUESTED** [review]

File: onadata/apps/main/models/meta_data.py [hunk 2]
```diff
@@ -129,66 +133,113 @@ class MetaData(models.Model):
...
+    data_filename = models.CharField(max_length=255, blank=True, null=True)
+    date_created = models.DateTimeField(default=timezone.now)
+    date_modified = models.DateTimeField(default=timezone.now)
 
     class Meta:
         app_label = 'main'
         unique_together = ('xform', 'data_type', 'data_value')
 
+    @property
+    def is_paired_data(self) -> bool:
```
[the remark this answers is not in the document]
Comment (jnm) [reply onadata/apps/main/models/meta_data.py:145]: i think i meant `data_type` :monocle_face:

File: onadata/libs/serializers/xform_serializer.py [hunk 3]
```diff
@@ -173,4 +173,4 @@ def get_url(self, obj):
 
     @check_obj
     def get_hash(self, obj):
-        return "%s" % (obj.file_hash or 'md5:')
+        return "%s" % (obj.hash or 'md5:')
```
Comment (jnm) [inline onadata/libs/serializers/xform_serializer.py:176]: let's avoid the name `hash` since it's also the name of a built-in function

**Review by jnm — CHANGES_REQUESTED** [review]

File: onadata/apps/main/models/meta_data.py [hunk 4]
```diff
@@ -129,66 +133,113 @@ class MetaData(models.Model):
...
-        if self.file_hash:
-            return self.file_hash
+    def _set_hash(self) -> str:
+        """
+        Sets `self.file_hash` if it is empty and returns it if no errors occur.
+        Otherwise, it returns an empty string
 
-        if not self.data_file:
-            return None
+        Notes: KoBoCAT returns an empty string is the object is a URL, but KPI
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply onadata/apps/main/models/meta_data.py:216]: Resolved. The comment has been removed by https://github.com/kobotoolbox/kobocat/commit/76d53d76fff271faa61fe5bc37fb476cd5f06f2d

File: onadata/libs/utils/hash.py (hunk 1 above)
Comment (noliveleger) [reply onadata/libs/utils/hash.py:75]: Would you mind to remind me what you meant exactly? 
Do you want to add this in the docstrings ?

**Pull request merged**